apps/system/js/devtools/logshake.js

```diff
@@ -122,52 +156,106 @@
       30: 'logsSDCardMaybeShared' // errno: filesystem ro-mounted
     },
 
-    handleCaptureLogsError: function(event) {
-      debug('handling capture logs error');
-      var error = '';
-      if (event) {
-        error = event.detail.error;
+    formatError: function(error) {
+      if (typeof error === 'string') {
+        return error;
       }
 
-      var moreInfos;
       if (typeof error === 'object') {
-        // Small heuristic for some frequent unix error cases
-        if ('operation' in error && 'unixErrno' in error) {
-          var errno = error.unixErrno;
-          debug('errno: ' + errno);
-
-          // Gracefully fallback to a generic error messages if we don't know
-          // this errno code.
-          if (!this.ERRNO_TO_MSG[errno]) {
-            errno = 0;
-          }
-
-          error = navigator.mozL10n.get('logsOperationFailed',
-                                        { operation: error.operation });
-          moreInfos = (function() {
-            ModalDialog.alert('logsSaveError',
-                              this.ERRNO_TO_MSG[errno], { title: 'ok' });
-          }).bind(this);
+        if ('operation' in error) {
+          return navigator.mozL10n.get('logsOperationFailed',
+                                       { operation: error.operation });
         }
       }
 
-      this._notify('logsSaveError', error, moreInfos);
+      return '';
     },
 
-    _notify: function(titleId, body, onclick) {
+    showErrorMessage: function(error, notif) {
+      if (notif) {
+        notif.close();
+      }
+
+      // Do nothing for error string
+      if (typeof error === 'string') {
+        return;
+      }
+
+      if (typeof error !== 'object') {
+        console.warn('Unexpected error type: ' + typeof error);
+        return;
+      }
+
+      // Small heuristic for some frequent unix error cases
+      if ('unixErrno' in error) {
+        var errno = error.unixErrno;
+        debug('errno: ' + errno);
+
+        // Gracefully fallback to a generic error messages if we don't know
+        // this errno code.
+        if (!this.ERRNO_TO_MSG[errno]) {
+          errno = 0;
+        }
+
+        ModalDialog.alert('logsSaveError',
+                          this.ERRNO_TO_MSG[errno], { title: 'ok' });
+      }
+    },
+
+    _notify: function(titleId, body, onclick, dataPayload) {
       var title = navigator.mozL10n.get(titleId) || titleId;
-      var notification =
-        new window.Notification(title, {body: body, tag: 'logshake'});
+      var payload = {
+        body: body,
+        tag: 'logshake',
+        data: {
+          systemMessageTarget: 'logshake',
+          logshakePayload: dataPayload || undefined
+        }
+      };
+      var notification = new Notification(title, payload);
       if (onclick) {
-        notification.onclick = onclick;
+        notification.onclick = onclick.bind(this, notification);
```

Comment:
Note: Binding a second time works for adding arguments, but it can't change the already bound `this` (here you could write `onclick.bind(null, notification)` and it would have the same effect).